### DISCUSSION TITLE
The correct way for adding products and product variations?

### DISCUSSION THREAD

**Discussion by omadawnx** [XiaoFaye/WooCommerce.NET#671] [Q&A] [answered]
First, I would like to thank you for your work, XiaoFaye!
I'm at a point where I can add categories, add products with images etc (all works great) -
But I'm struggling with Product Variations, it's running but the variations just don't show on the WC admin.

This is the code I'm currently testing with:
In this case, I'm trying to add variation products, each variation has a SKU and a description:

`
            // a category for the product (previously created)
            ProductCategoryLine productCategoryLine = new ProductCategoryLine()
            {
                id = 118,
                name = "test",
                slug = "test"
            };

            // prepare a sample product image
            ProductImage productImage = new ProductImage()
            {
                name = "test",
                src = "https://www.2sit.co.il/uploaded_files/images/tak_30_U5018.jpg",
                position = 0
            };

            // prepare an attribute for variants and a variant
            VariationAttribute varattr = new VariationAttribute() { name = "SKU" };
            Variation variation1 = new Variation()
            {
                sku = "0030303030",
                description = "test description"
            };
            // another one for fun
            Variation variation2 = new Variation()
            {
                sku = "0023232323",
                description = "test description 2"
            };

            // preparing the main product:
            Product product = new Product()
            {
                type = "variable",
                name = "test",
                description = "desciption",
                short_description = "short desc",
                price = 0,
                categories = new List<ProductCategoryLine>() { productCategoryLine },
                images = new List<ProductImage>() { productImage },
                attributes = new List<ProductAttributeLine>() {     // this is probably wrong?
                    new ProductAttributeLine
                    {
                        name = "SKU",
                        position = 1,
                        variation = true,
                        visible = true
                    }
                }
            };
            Product newprod = await wc.Product.Add(product);    // adding the product (this works but no trace for the attributes)

            Variation newvar = await wc.Product.Variations.Add(variation1, (int)newprod.id);    // adding first var. returns a new id, but doesnt show on backend
            newvar = await wc.Product.Variations.Add(variation2, (int)newprod.id);              // adding second, same thing
` 
However, the variations tab of this new product (which otherwise uploaded perfectly), shows this:
![image](https://user-images.githubusercontent.com/31862808/146380341-9d8ef244-50d4-4004-9635-9a9161bd29ba.png)

I would very much appreciate a clear guide on how to do this, thank you for any help!

**Comment by XiaoFaye** [comment]
Have you created attributes in the system?

**Reply by omadawnx** [reply]
Yes I did, but now I understand I'm probably thinking about "variations" in a wrong way..
There is a system attribute "SKU" but it had no "options" (or "items"). I think the system wants me to add all these options in advance (like color: red, green, blue)..  when adding custom options, the variations finally appear.
But I cant figure out this attribute system, maybe because I'm sick and cant even think straight :(

I'll need to figure out the story about these variation thingies in WC..
Is there anyway you can guide me on how to "prepare" WC to receive variations, each variation has a SKU and a description?

I'll take the time later to try to understand WC, but it will have to wait until my head can work again.

Thank you vey much!

**Comment by omadawnx** [comment, marked as answer]
ok.. after sleeping on it, finally got it to work.
You were right about the attributes, they are like an axis for hanging variations on.
So I needed to create a product attribute with a value (options) of "something" (some random value, in this case 00000000),
and when creating the variations, they need to be attached to that attribute with the specific option.

  here's a working code I'm using to test this:

`

           // a category for the product (previously created)
            ProductCategoryLine productCategoryLine = new ProductCategoryLine()
            {
                id = 118,
                name = "test",
                slug = "test"
            };

            // prepare a sample product image
            ProductImage productImage = new ProductImage()
            {
                name = "test",
                src = "https://www.2sit.co.il/uploaded_files/images/tak_30_U5018.jpg",
                position = 0
            };

            // preparing the main product:
            Product product = new Product()
            {
                type = "variable",
                name = "test",
                description = "desciption",
                short_description = "short desc",
                price = 0,
                categories = new List<ProductCategoryLine>() { productCategoryLine },
                images = new List<ProductImage>() { productImage },
                attributes = new List<ProductAttributeLine>() {     // add attributes for variations
                    new ProductAttributeLine
                    {
                        name = "SKU",
                        position = 1,
                        variation = true,
                        visible = true,options = new List<string>() { "00000000" }
                    }
                }
            };

            // add variations and attach them to a specific option of an exisiting product attribute
            VariationAttribute varattr = new VariationAttribute() { name = "SKU", option = "00000000" };
            Variation variation1 = new Variation()
            {
                sku = "111111111x",
                description = "test description 1",
                attributes = new List<VariationAttribute>() { varattr }
            };
            // another one for fun
            Variation variation2 = new Variation()
            {
                sku = "222222222y",
                description = "test description 2",
                attributes = new List<VariationAttribute>() { varattr }
            };


            Product newprod = await wc.Product.Add(product);    // adding the product 

            Variation newvar = await wc.Product.Variations.Add(variation1, (int)newprod.id);    // adding first var 
            newvar = await wc.Product.Variations.Add(variation2, (int)newprod.id);              // adding second var


`

So thank you for making WooCommerce.NET, keep up the good work!
